### PR TITLE
test/cql-pytest: run tests with tablets

### PR DESCRIPTION
**Pull request by denesb**
Add `--experimental-features=tablets` to both `test/cql-pytest/suite.yaml` and `test/cql-pytest/run.py`, so tablets are enabled. Detect tablet support in `contest.py` and add an xfail and skip marker to mark tests that fail/crash with tablets. These are expected to be fixed soon.

Some tests checking things around alter-keyspace, had to force-disable tablets on the created keyspace, because tablets interfere with the test (a keyspace with tablets cannot have simple strategy for example).
Tablets were also interfering with `test_keyspace.py:test_storage_options_local`, because it is expecting `system_schema.scylla_keyspaces` to not have any entries for local storage keyspace, but they have it if tablets are enabled. Adjust the test to account for this.

### CODE REVIEW

**Comment by denesb** [comment]
This is a replacement of https://github.com/scylladb/scylladb/pull/16721, dropping the controversial running-each-test-twice part. This PR now just focuses on making cql-pytest work with tablets enabled.

**Comment by denesb** [comment]
> [topology_experimental_raft.test_tablets.dev.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5891/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_tablets_dev_5) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.dev.5)

https://github.com/scylladb/scylladb/issues/14746

File: test/cql-pytest/util.py [hunk 1]
```diff
@@ -63,15 +63,38 @@ def format_tuples(tuples=None, **kwargs):
...
 
+def keyspace_has_tablets(cql, keyspace):
+    """ Return true if the test_keyspace was created with tablets.
+
+    We support running cql-pytest against an older version of scylla, so we do
+    the detection in a way that accounts for scylla possibly not even knowing
+    what tablets is.
+
+    If the keyspace was created with tablets, it will have an entry in
+    `system_schema.scylla_keyspaces`, with `initial_tablets` set to 0.
```
Comment (nyh) [inline test/cql-pytest/util.py:74]: As the code below does, you only need to check if it exists - it doesn't necessarily need to be 0 (even though 0 is indeed the default, maybe it was chosen explicitly).

File: test/cql-pytest/cassandra_tests/validation/operations/alter_test.py [hunk 2]
```diff
@@ -206,8 +206,8 @@ def testCreateAlterKeyspaces(cql, test_keyspace, this_dc):
 # Reproduces CASSANDRA-12681 and Scylla #10036
 def testAlterKeyspaceWithNoOptionThrowsConfigurationException(cql, test_keyspace, this_dc):
     # Create keyspaces
-    with create_keyspace(cql, "replication={ 'class' : 'NetworkTopologyStrategy', '" + this_dc + "' : 3 }") as abc:
-        with create_keyspace(cql, "replication={ 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 3 }") as xyz:
+    with create_keyspace(cql, "replication={ 'class' : 'NetworkTopologyStrategy', '" + this_dc + "' : 3 }", force_no_tablets=True) as abc:
+        with create_keyspace(cql, "replication={ 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 3 }", force_no_tablets=True) as xyz:
```
Comment (nyh) [inline test/cql-pytest/cassandra_tests/validation/operations/alter_test.py:210]: In my PR, I made this test fails_with_tablets - because it demonstrates a real difference between what Cassandra allows, and Scylla used to allow and no longer does.
But I won't insist. I also wasn't sure what exactly to do about this test.

File: test/cql-pytest/conftest.py [hunk 3]
```diff
@@ -221,3 +221,18 @@ def temp_workdir():
...
+
+@pytest.fixture(scope="function")
+def xfail_tablets(request, has_tablets):
+    if has_tablets:
+        request.node.add_marker(pytest.mark.xfail(reason='Test expected to fail with tablets experimental feature on'))
+
+@pytest.fixture(scope="function")
+def skip_with_tablets(request, has_tablets):
+    if has_tablets:
+        pytest.skip("test may crash with tablets enabled")
```
Comment (nyh) [inline test/cql-pytest/conftest.py:238]: Nice catch. In my own tests I only used test/cql-pytest/run, not test.py, so I didn't use that horrible "on-internal-error-causes-a-crash" mode I didn't see the need for skip. But yes, in test.py it's unforunately needed (https://github.com/scylladb/scylladb/issues/13352).

File: test/cql-pytest/suite.yaml [hunk 4]
```diff
@@ -2,3 +2,9 @@ type: Python
 pool_size: 4
 dirties_cluster:
   - test_native_transport
+extra_scylla_cmdline_options:
+  - '--experimental-features=udf'
+  - '--experimental-features=consistent-topology-changes'
+  - '--experimental-features=broadcast-tables'
```
Comment (nyh) [inline test/cql-pytest/suite.yaml:8]: Why did you add this "broadcast tables" here? Do we test them in cql-pytest?

File: test/cql-pytest/test_keyspace.py [hunk 5]
```diff
@@ -207,7 +207,7 @@ def creates(count):
 def test_storage_options_local(cql, scylla_only):
     ksdef = "WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : '1' } " \
             "AND STORAGE = { 'type' : 'LOCAL' }"
-    with new_test_keyspace(cql, ksdef) as keyspace:
+    with new_test_keyspace(cql, ksdef, force_no_tablets=True) as keyspace:
```
Comment (nyh) [inline test/cql-pytest/test_keyspace.py:210]: This is sad, but not your fault. This test is faulty, it assumes that the existance of scylla_keyspaces is what determines if there's local storage or not, which is wrong.
I admit that making it not use tablets is easier than fixing this test.
We'll probably run into this problematic test again in the future, when we add more stuff into scylla_keyspaces, and will then need to fix this test for real :-(

**Comment by avikivity** [comment]
I think I made a mistake when I wanted all tests to run in dual mode. The cql-level tests don't benefit from it.

SELECT mutation_fragments() does, and probably others too, but perhaps we can annotate the tests to declare whether they depend on replication or not.

**Comment by nyh** [comment]
> SELECT mutation_fragments() does, and probably others too, but perhaps we can annotate the tests to declare whether they depend on replication or not.

Note that we can quite easily do what Botond did in the previous PR but just for specific tests: We can using pytest's nice "parametrize" feature you can run a certain test twice, once with tablets and once without tablets. You just don't need to do it for every test (e.g., no need to do it for test for CQL syntax, etc.). 

You can see an example of the "parametrize" usage for another feature in test/cql-pytest/test_secondary_index.py - there are a few tests where I want to show that a certain filtering query yields correct results whether or not a secondary index is used - so I used `@pytest.mark.parametrize("use_index", ...` to have pytest run this test twice - once with a secondary index and once without. Both versions should pass (and as you can guess, the index one fails ;-)).

File: test/cql-pytest/util.py [hunk 6]
```diff
@@ -63,15 +63,38 @@ def format_tuples(tuples=None, **kwargs):
...
+    So here, we simply query this table, looking for a partition for the
+    appropriate keyspace. If the result has the `initial_tablets` column and it
+    is set, the keyspace has tablets.
+    """
+    # Need to use network strategy, otherwise tablets will not be enabled.
+    res = list(cql.execute(f"SELECT * FROM system_schema.scylla_keyspaces WHERE keyspace_name='{keyspace}'"))
+    if not res:
+        return False
+    return getattr(res[0], "initial_tablets", None) is not None
```
[the remark this answers is not in the document]
Comment (denesb) [reply test/cql-pytest/util.py:83]: Yes, but here I want to distinguish the case of the attribute existing, but being set to null (None). I will add a comment.
The keyspace might have a row here, due to storage options, but has the tablets-related fields set to null.

File: test/cql-pytest/util.py (hunk 1 above)
Comment (denesb) [reply test/cql-pytest/util.py:74]: Yes, will fix.

File: test/cql-pytest/cassandra_tests/validation/operations/alter_test.py (hunk 2 above)
Comment (denesb) [reply test/cql-pytest/cassandra_tests/validation/operations/alter_test.py:210]: Yes, and any test which works only with tablets disabled, has an expiry date on it, as we want to retire vnodes at one point in the future. So any such workarounds, are temporary by nature.

File: test/cql-pytest/suite.yaml (hunk 4 above)
Comment (denesb) [reply test/cql-pytest/suite.yaml:8]: Copy pasta, will remove.

File: test/cql-pytest/test_keyspace.py (hunk 5 above)
Comment (denesb) [reply test/cql-pytest/test_keyspace.py:210]: Fixing this is so simple, I might just do it in v2, instead of using this marker here and explaining why I did it in the patch notes.

**Comment by denesb** [comment]
New in v2:
* Split into multiple patches for easier reviewability
* Cherry-pick @nyh's fix for test_tablets.py
* Proper fix for test_keyspace.py::test_storage_options_local()
* Move tablet disable to callers (only 3 remain)
* Fix `keyspace_has_tablets()` to also work for cassandra
* Various comment improvements in util.py and conftest.py

**Comment by nyh** [comment]
One quick comment after a brief look: It seems that because your "hastablets" function uses scylla_only, I think all tests with xfail_tablets will be skipped on Cassandra, and that's unnecessary.
In my PR the tablets checking function simply returns false on cassanda - it only returns true on Scylla with tablets.

File: test/cql-pytest/test_keyspace.py (hunk 5 above)
Comment (denesb) [reply test/cql-pytest/test_keyspace.py]: Fixed in v2 (properly).

File: test/cql-pytest/suite.yaml (hunk 4 above)
Comment (denesb) [reply test/cql-pytest/suite.yaml]: Removed in v2.

File: test/cql-pytest/conftest.py (hunk 3 above)
Comment (denesb) [reply test/cql-pytest/conftest.py]: These are all in the select_from_mutation_fragments.py, we can remove this fixture altogether once https://github.com/scylladb/scylladb/pull/16802 gets in.

File: test/cql-pytest/util.py [hunk 7]
```diff
@@ -63,15 +63,38 @@ def format_tuples(tuples=None, **kwargs):
...
+# The force_no_tablets options allows disabling tablets, if scylla would enable
+# it by default.
 @contextmanager
-def new_test_keyspace(cql, opts):
+def new_test_keyspace(cql, opts, force_no_tablets=False):
     keyspace = unique_name()
     cql.execute("CREATE KEYSPACE " + keyspace + " " + opts)
+    if force_no_tablets and keyspace_has_tablets(cql, keyspace):
+        cql.execute("DROP KEYSPACE " + keyspace)
+        cql.execute("CREATE KEYSPACE " + keyspace + " " + opts + " AND TABLETS = {'enabled': false}")
```
[the remark this answers is not in the document]
Comment (denesb) [reply test/cql-pytest/util.py]: Since in v2 there were only 3 callers left, for 2 `create_keyspace_*()` variants, I just extracted the logic into the callers instead. We shouldn't make it easy to disable tablets in a test.

File: test/cql-pytest/util.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (denesb) [reply test/cql-pytest/util.py:98]: Added a comment in v2.

File: test/cql-pytest/util.py (hunk 1 above)
Comment (denesb) [reply test/cql-pytest/util.py]: Fixed in v2.

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
Thanks, pushed.

I will want to prepare on sunday a followup patches a --vnodes option to test/{cql-pytest,alternator}/run, for a developer optionally running the cql-pytests **without** tablets (yes, I know this option won't be interesting for long, but I think for the near future, we may sometimes want to try things with it).

I am also curious how much adding tablets slowed down cql-pytest and alternator tests, but I'm still waiting forever for Scylla to compile so I'll know later...
